File: example/java/src/mainGenerator/java/io/heartpattern/gpg/example/java/generator/ExampleGenerator.java
package io.heartpattern.gpg.example.java.generator;

import com.squareup.javapoet.CodeBlock;
import com.squareup.javapoet.JavaFile;
import com.squareup.javapoet.MethodSpec;
import com.squareup.javapoet.TypeSpec;
import io.heartpattern.gcg.api.Generator;
import io.heartpattern.gcg.api.java.JavaCodeGenerator;

import javax.lang.model.element.Modifier;
import java.util.Collection;
import java.util.Collections;

@Generator
public class ExampleGenerator implements JavaCodeGenerator {
    @Override
    public Collection<JavaFile> generateJava() {
        return Collections.singleton(
                JavaFile.builder(
                        "io.heartpattern.gpg.example.java",
                        TypeSpec.classBuilder("GeneratedClass")
                                .addModifiers(Modifier.PUBLIC, Modifier.FINAL)
                                .addMethod(
                                        MethodSpec.methodBuilder("printHello")
                                                .addModifiers(Modifier.PUBLIC, Modifier.FINAL)
                                                .returns(void.class)
                                                .addParameter(String.class, "name")
                                                .addStatement(
                                                        "UsedInGeneratedCode.printHelloDelegate(name)"
                                                )
                                                .build()
                                )
                                .build()
                ).build()
        );
    }
}

File: example/kotlin/src/mainGenerator/kotlin/io/heartpattern/gcg/example/kotlin/generator/ExampleGenerator.kt
package io.heartpattern.gcg.example.kotlin.generator

import com.squareup.kotlinpoet.FileSpec
import com.squareup.kotlinpoet.FunSpec
import io.heartpattern.gcg.api.Generator
import io.heartpattern.gcg.api.kotlin.KotlinCodeGenerator

@Generator
class ExampleGenerator : KotlinCodeGenerator {
    override fun generateKotlin(): Collection<FileSpec> {
        return listOf(
            FileSpec.builder("io.heartpattern.gcg.example.kotlin", "GeneratedCode")
                .addFunction(
                    FunSpec.builder("printHello")
                        .addParameter("name", String::class)
                        .addStatement("printHelloDelegate(name)")
                        .build()
                )
                .build()
        )
    }
}